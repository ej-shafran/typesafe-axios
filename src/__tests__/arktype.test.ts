import { taxios } from "..";

import { scope } from "arktype";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = scope({
  follower: {
    id: "string",
    "role?": '"admin" | "guest"',
  },
  user: {
    id: "string",
    username: "string",
    email: "string",
    address: {
      street: "string",
      "zipcode?": "string",
    },
    followers: "follower[]",
  },
}).compile().user;

type TYPE = typeof SCHEMA.infer;
function ensure(_: TYPE) { }

describe("ArkType interop", () => {
  it("passes along data with the correct type", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);
    const { data } = await taxios(SCHEMA).get(UTILS.API);

    expect(data).toEqual(UTILS.VALID);
    ensure(data);
  });

  it("throws a validation error when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    try {
      await taxios(SCHEMA).get(UTILS.API);
    } catch (error) {
      expect(JSON.stringify(error, null, 2)).toMatchSnapshot();
    }
  });
});
