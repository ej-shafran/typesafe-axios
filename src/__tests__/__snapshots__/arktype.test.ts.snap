// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArkType interop throws a validation error when the data is invalid 1`] = `
"{
  "cause": [
    {
      "code": "missing",
      "path": [
        "email"
      ],
      "writers": {}
    },
    {
      "code": "domain",
      "path": [
        "address"
      ],
      "data": "ABC123 St.",
      "source": "object",
      "writers": {}
    }
  ]
}"
`;
